java/src/main/java/com/vmware/App.java

```diff
@@ -0,0 +1,88 @@
+ */
+public class App {
+
+    private final static Logger logger = LoggerFactory.getLogger(App.class);
+    public static final String OTEL_RESOURCE_ATTRIBUTES_KEY = "otel.resource.attributes";
+    public static final String OTEL_RESOURCE_ATTRIBUTES_VALUE = "service.name=HelloWorldJavaService";
+    public static final String INSTRUMENTATION_LIBRARY_NAME = "instrumentation-library-name";
+    public static final String INSTRUMENTATION_VERSION = "1.0.0";
+    static Tracer tracer;
+
+    public static void main(String[] args) throws InterruptedException {
+
+        /*this will make sure that a proper service.name attribute is set on all the
+         spans/metrics.*/
+        System.setProperty(OTEL_RESOURCE_ATTRIBUTES_KEY, OTEL_RESOURCE_ATTRIBUTES_VALUE);
+
+        /*tracer must be acquired, which is responsible for creating spans and interacting with the Context*/
+        tracer = getTracer();
+
+        /*an automated way to propagate the parent span on the current thread*/
+        for (int index = 0; index < 3; index++) {
+            /*create a span by specifying the name of the span. The start and end time of the span is automatically set by the OpenTelemetry SDK*/
+            Span parentSpan = tracer.spanBuilder("parentSpan").startSpan();
+            logger.info("In parent method. TraceID : {}", parentSpan.getSpanContext().getTraceIdAsHexString());
+
+            /*put the span into the current Context*/
+            try (Scope scope = parentSpan.makeCurrent()) {
+
+                /*annotate the span with attributes specific to the represented operation, to provide additional context*/
+                parentSpan.setAttribute("parentIndex", index);
+                childMethod(parentSpan);
+            } catch (Throwable throwable) {
+                parentSpan.setStatus(StatusCode.ERROR, "Something wrong with the parent span");
+            } finally {
+                /*closing the scope does not end the span, this has to be done manually*/
+                parentSpan.end();
+            }
+        }
+
+        /*sleep for a bit to let everything settle*/
+        Thread.sleep(2000);
+    }
+
+    private static void childMethod(Span parentSpan) {
+
+        tracer = getTracer();
+
+        /*setParent(...) is not required, `Span.current()` is automatically added as the parent*/
+        Span childSpan = tracer.spanBuilder("childSpan").setParent(Context.current().with(parentSpan))
+                .startSpan();
+        logger.info("In child method. TraceID : {}", childSpan.getSpanContext().getTraceIdAsHexString());
+
+        /*put the span into the current Context*/
+        try (Scope scope = childSpan.makeCurrent()) {
+            Thread.sleep(1000);
+        } catch (Throwable throwable) {
+            childSpan.setStatus(StatusCode.ERROR, "Something wrong with the child span");
+        } finally {
+            childSpan.end();
+        }
+    }
+
+    private static Tracer getTracer() {
+        if (tracer == null) {
```

Comment:
This may not work properly in a multi threaded environment.  At the very least, declare the `tracer` variable as volatile.  Even declaring `tracer` as volatile won't ensure that you get one tracer because of race conditions.  Golang has a sync.Once type that guarantees that it will perform some initialization exactly once.  Maybe JAVA has a similar construct?

Reply:
Yes, in Java we can obtain a lock by synchronization technique. I am updating the code. That will make it threadsafe.